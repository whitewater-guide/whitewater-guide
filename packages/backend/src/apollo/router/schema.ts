import { GraphQLSchema } from 'graphql';
import { makeExecutableSchema } from 'graphql-tools';
import { fileLoader, mergeTypes } from 'merge-graphql-schemas';
import { join } from 'path';
import log from '../../log/index';
import { resolvers } from './resolvers';

const logger = log.child({ module: 'apollo' });

let typeDefs: string;
let schema: GraphQLSchema;

async function loadSchema() {
  const typesArray = fileLoader(join(process.cwd(), 'src'), { recursive: true, extensions: ['.graphql'] });
  typeDefs = mergeTypes(typesArray);
  schema = makeExecutableSchema({
    typeDefs,
    resolvers,
    allowUndefinedInResolve: false,
  });
  logger.info('Initialized GRAPHQL schema');
  return schema;
}

export async function getSchema(): Promise<GraphQLSchema> {
  if (schema) {
    return Promise.resolve(schema);
  }
  return loadSchema();
}

export async function getTypeDefs(): Promise<string> {
  if (typeDefs) {
    return Promise.resolve(typeDefs);
  }
  await loadSchema();
  return typeDefs;
}
