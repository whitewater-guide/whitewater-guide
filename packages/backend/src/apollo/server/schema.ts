import { makeExecutableSchema } from 'apollo-server-koa';
import { GraphQLSchema } from 'graphql';
import { fileLoader, mergeTypes } from 'merge-graphql-schemas';
import { join } from 'path';

import { AdminDirective } from '../directives';
import { logger } from '../logger';
import { resolvers } from './resolvers';

let typeDefs: string;
let schema: GraphQLSchema;

async function loadSchema() {
  const typesArray = fileLoader(
    join(process.cwd(), process.env.NODE_ENV === 'production' ? 'dist' : 'src'),
    {
      recursive: true,
      extensions: ['.graphql'],
    },
  );
  // it seems that definitions are incorrect and it still returns string
  typeDefs = mergeTypes(typesArray) as any;
  const result = makeExecutableSchema({
    typeDefs,
    resolvers,
    allowUndefinedInResolve: false,
    schemaDirectives: {
      admin: AdminDirective as any,
    },
  });
  logger.info('Initialized GRAPHQL schema');
  return result;
}

export async function getSchema(): Promise<GraphQLSchema> {
  if (!schema) {
    schema = await loadSchema();
  }
  return schema;
}

export async function getTypeDefs(): Promise<string> {
  if (typeDefs) {
    return Promise.resolve(typeDefs);
  }
  await loadSchema();
  return typeDefs;
}
