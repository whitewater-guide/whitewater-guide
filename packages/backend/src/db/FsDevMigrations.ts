// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import { FsMigrations } from 'knex/lib/migrate/sources/fs-migrations';

// When restoring production dump on development, migration names will have '.js' extension
// Therefore, development migrations with '.ts' will conflict them
// this is the workaround
// https://github.com/knex/knex/blob/master/lib/migrate/sources/fs-migrations.js
export class FsDevMigrations extends FsMigrations {
  // eslint-disable-next-line @typescript-eslint/no-useless-constructor
  constructor(
    migrationDirectories: string | string[],
    sortDirsSeparately?: boolean,
    loadExtensions?: string[],
  ) {
    super(migrationDirectories, sortDirsSeparately, loadExtensions);
  }

  getMigrationName(migration: any): string {
    return migration.file.replace('.ts', '.js');
  }
}
