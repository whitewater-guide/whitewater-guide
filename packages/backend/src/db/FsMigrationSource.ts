import fs from 'node:fs/promises';
import path from 'node:path';

import type { Knex } from 'knex';

// https://knexjs.org/guide/migrations.html#custom-migration-sources
export default class FsMigrationSource implements Knex.MigrationSource<string> {
  // eslint-disable-next-line node/no-process-env
  private readonly ext = process.env.NODE_ENV === 'production' ? 'js' : 'ts';
  private readonly directory: string;

  constructor(directory: string) {
    this.directory = directory;
  }
  // Must return a Promise containing a list of migrations.
  // Migrations can be whatever you want,
  // they will be passed as arguments to getMigrationName
  // and getMigration
  //
  // In our case migrations are filenames without extension
  async getMigrations(): Promise<string[]> {
    // In this example we are just returning migration names
    const dirContent = await fs.readdir(this.directory);
    // assume that migrations have extensions (== no directories with '.' symbol)
    return dirContent
      .filter((f) => f.includes('.'))
      .map((f) => f.split('.')[0])
      .sort();
  }

  // When restoring production dump on development, migration names will have '.js' extension
  getMigrationName(migration: string): string {
    return migration + '.js';
  }

  async getMigration(migration: string): Promise<Knex.Migration> {
    const moduleName = path.resolve(
      process.cwd(),
      this.directory,
      `${migration}.${this.ext}`,
    );
    return import(moduleName);
  }
}
