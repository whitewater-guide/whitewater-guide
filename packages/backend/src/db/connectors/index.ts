export * from './createConnectors';
export * from './OffsetConnector';
export * from './types';
