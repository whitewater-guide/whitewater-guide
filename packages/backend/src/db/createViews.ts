import { readFileSync } from 'fs';
import Knex from 'knex';
import path from 'path';

export async function createView(db: Knex, view: string, version = 1) {
  for (let i = version; i >= 1; i--) {
    const file = path.resolve(
      __dirname,
      '../migrations',
      `${i.toString(10).padStart(3, '0')}/${view}_view.sql`,
    );
    let sql: string;
    try {
      sql = readFileSync(file).toString();
    } catch (e) {
      continue;
    }
    await db.schema.raw(sql);
    return;
  }
  throw new Error(
    `No sql files for view '${view}' of version ${version} or lower found`,
  );
}

// Views will be created in reverse order
export async function createViews(
  db: Knex,
  version: number,
  ...views: string[]
) {
  const inverted = [...views].reverse();
  for (const view of inverted) {
    await createView(db, view, version);
  }
}
