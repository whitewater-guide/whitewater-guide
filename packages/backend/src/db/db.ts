/* eslint-disable node/no-process-env */
import Knex from 'knex';

import knexConfig from './knexfile';

const env = process.env.NODE_ENV || 'development';
const config: Knex.Config = knexConfig[env];

export const knex = Knex(config);

let testTransaction: Knex.Transaction | null = null;

export function db(getKnexInstance = false): Knex<any, any> {
  if (env === 'test' && !getKnexInstance) {
    if (testTransaction) {
      return testTransaction;
    }
    throw new Error('No saved test transaction found');
  }
  return knex;
}

// See here https://github.com/tgriesser/knex/issues/2076
// And here https://gist.github.com/odigity/7f37077de74964051d45c4ca80ec3250
export const holdTransaction = () => {
  if (env !== 'test') {
    throw new Error('holdTransaction is only available in test mode');
  }
  const tmp: any = {};
  const p = new Promise((resolve) => {
    tmp.resolve = resolve;
  });
  knex
    .transaction((tx) => {
      testTransaction = tx;
      tmp.resolve();
    })
    .catch(() => {
      // ignore
    });
  return p;
};

export const rollbackTransaction = async () => {
  if (env !== 'test') {
    throw new Error('rollbackTransaction is only available in test mode');
  }
  if (testTransaction) {
    await testTransaction.rollback();
    testTransaction = null;
  }
};
