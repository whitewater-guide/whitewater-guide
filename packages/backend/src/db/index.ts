import db from './db';

export default db;
export * from './updatedAtTrigger';
export * from './runSqlFile';
export * from './createViews';
export * from './dropViews';
export * from './rawUpsert';
export * from './types';
export * from './db';
export * from './queryBuilders';
