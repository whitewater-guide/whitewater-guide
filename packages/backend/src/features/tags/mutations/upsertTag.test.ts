import { ApolloErrorCodes } from '@whitewater-guide/commons';
import type { TagInput } from '@whitewater-guide/schema';
import { TagCategory } from '@whitewater-guide/schema';
import { gql } from 'graphql-tag';

import { db, holdTransaction, rollbackTransaction } from '../../../db/index';
import { ADMIN, EDITOR_GA_EC, TEST_USER } from '../../../seeds/test/01_users';
import { anonContext, fakeContext } from '../../../test/index';
import type {
  UpsertTagMutation,
  UpsertTagMutationResult,
} from './upsertTag.test.generated';
import { testUpsertTag } from './upsertTag.test.generated';

beforeEach(holdTransaction);
afterEach(rollbackTransaction);

const _mutation = gql`
  mutation upsertTag($tag: TagInput!) {
    upsertTag(tag: $tag) {
      id
      name
      category
    }
  }
`;

const tag: TagInput = {
  id: 'new_tag',
  name: 'New tag',
  category: TagCategory.Misc,
};

describe('resolvers chain', () => {
  it('anon should not pass', async () => {
    const result = await testUpsertTag({ tag }, anonContext());
    expect(result).toHaveGraphqlError(ApolloErrorCodes.UNAUTHENTICATED);
  });

  it('user should not pass', async () => {
    const result = await testUpsertTag({ tag }, fakeContext(TEST_USER));
    expect(result).toHaveGraphqlError(ApolloErrorCodes.FORBIDDEN);
  });

  it('editor should not pass', async () => {
    const result = await testUpsertTag({ tag }, fakeContext(EDITOR_GA_EC));
    expect(result).toHaveGraphqlError(ApolloErrorCodes.FORBIDDEN);
  });

  it('should throw on invalid input', async () => {
    const invalidInput = {
      id: 'a b',
      name: 'x',
      category: 'misc', // Bad category is validated by Graphql enum
    };
    const result = await testUpsertTag(
      // @ts-expect-error: invalid input
      { tag: invalidInput },
      fakeContext(ADMIN),
    );
    expect(result).toHaveGraphqlValidationError();
  });
});

describe('insert', () => {
  it('should return result', async () => {
    const result = await testUpsertTag({ tag }, fakeContext(ADMIN));
    expect(result.errors).toBeUndefined();
    expect(result.data?.upsertTag).toEqual({ ...tag });
  });

  it('should add one more tag', async () => {
    const { count } = await db().table('tags').count().first();
    const initialTagsCount = Number(count);
    await testUpsertTag({ tag }, fakeContext(ADMIN));
    const { count: tagCount } = await db().table('tags').count().first();
    expect(Number(tagCount) - initialTagsCount).toBe(1);
  });
});

describe('update', () => {
  const input: TagInput = {
    id: 'waterfalls',
    name: 'Old waterfalls',
    category: TagCategory.Kayaking,
  };

  let initialTagsCount: number;
  let updateResult: UpsertTagMutationResult | null = null;
  let updatedTag: UpsertTagMutation['upsertTag'] | null = null;

  beforeAll(async () => {
    const { count } = await db(true).table('tags').count().first();
    initialTagsCount = Number(count);
  });

  beforeEach(async () => {
    updateResult = await testUpsertTag({ tag: input }, fakeContext(ADMIN));
    updatedTag = updateResult?.data?.upsertTag;
  });

  afterEach(() => {
    updateResult = null;
    updatedTag = null;
  });

  it('should return result', () => {
    expect(updateResult?.errors).toBeUndefined();
    expect(updateResult?.data).toBeDefined();
    expect(updatedTag).toBeDefined();
    expect(updatedTag?.id).toBe(input.id);
  });

  it('should not change total number of tags', async () => {
    const { count } = await db().table('tags').count().first();
    expect(initialTagsCount - Number(count)).toBe(0);
  });
});

describe('i18n', () => {
  const inputFr: TagInput = {
    id: 'waterfalls',
    name: 'Le waterfalls',
    category: TagCategory.Misc,
  };

  it('should add new translation', async () => {
    const { count: oldCount } = await db()
      .table('tags_translations')
      .count()
      .first();
    await testUpsertTag({ tag: inputFr }, fakeContext(ADMIN, 'fr'));
    const { count: newCount } = await db()
      .table('tags_translations')
      .count()
      .first();
    expect(Number(newCount) - Number(oldCount)).toBe(1);
    const { name } = await db()
      .table('tags_view')
      .select('name')
      .where({ language: 'fr', id: inputFr.id })
      .first();
    expect(name).toBe('Le waterfalls');
  });

  it('should modify common props in other language', async () => {
    await testUpsertTag({ tag: inputFr }, fakeContext(ADMIN, 'fr'));
    const { category } = await db()
      .table('tags_view')
      .select('category')
      .where({ language: 'en', id: inputFr.id })
      .first();
    expect(category).toBe(TagCategory.Misc);
  });

  it('should modify existing translation', async () => {
    const { count: oldCount } = await db()
      .table('tags_translations')
      .count()
      .first();
    await testUpsertTag({ tag: inputFr }, fakeContext(ADMIN, 'en'));
    const { count: newCount } = await db()
      .table('tags_translations')
      .count()
      .first();
    expect(newCount).toBe(oldCount);
    const { name } = await db()
      .table('tags_view')
      .select('name')
      .where({ language: 'en', id: inputFr.id })
      .first();
    expect(name).toBe('Le waterfalls');
  });
});

it('should sanitize input', async () => {
  const dirty = { ...tag, name: "it's a \\ $1 slash with . ?" };
  const result = await testUpsertTag({ tag: dirty }, fakeContext(ADMIN));
  expect(result).toHaveProperty(
    'data.upsertTag.name',
    "it's a \\ $1 slash with . ?",
  );
});
