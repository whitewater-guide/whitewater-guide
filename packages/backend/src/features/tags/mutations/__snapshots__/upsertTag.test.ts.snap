// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolvers chain should throw on invalid input 1`] = `
Object {
  "0": Object {
    "message": "\\"id\\" with value \\"a b\\" fails to match the required pattern: /^[0-9a-zA-Z_\\\\-]{3,64}$/",
    "path": "tag.id",
  },
  "1": Object {
    "message": "\\"name\\" length must be at least 3 characters long",
    "path": "tag.name",
  },
}
`;
